R/Quosure.R

```diff
@@ -0,0 +1,285 @@
+#' @export
+setMethod("get_var", signature("Quosure", "character"), function(object, var) {
+  get(var, envir = object@env)
+})
+
+#' @param x (`Quosure`)
+#' @param i (`character`) name of the binding in environment (name of the variable)
+#' @param j not used
+#' @param ... not used
+#' @rdname get_var
+#' @export
+setMethod("[[", c("Quosure", "ANY", "missing"), function(x, i, j, ...) {
+  get_var(x, i)
+})
+
+
+#' Get the code from the `Quosure`
+#'
+#' @param object (`Quosure`)
+setMethod("get_code", signature("Quosure"), function(object) {
+  object@code
+})
+
+#' Join two `Quosure` objects
+#'
+#' Combine two `Quosure` object by merging their environments and the code.
+#' Not every `Quosure` objects can be combined:
+#' - if their environments contains objects of the same name but not identical
+#' - if `object2` has unique code element placed before common element. This means that `object2`
+#' in the environment of the `object2` was evaluated some extra code before which can influence
+#' reproducibility
+#' - more cases todo
+#' @param object (`Quosure`)
+#' @param object2 (`Quosure`)
+#' @export
+setGeneric("join", function(object, object2) {
+  standardGeneric("join")
+})
+
+setMethod("join", signature("Quosure", "Quosure"), function(object, object2) {
+  # todo: revise chunks_merge_chunks and make robust join method
+
+  # object2 can't have modified object of the same name! See chunks_push_chunks
+  common_names <- intersect(ls(object@env), ls(object2@env))
+  is_overwritten <- vapply(common_names, function(x) {
+    !identical(get(x, object@env), get(x, object2@env))
+  }, logical(1))
+
+  # code
+  a_is_shared_code <- object@code %in% object2@code
+  b_is_shared_code <- object2@code %in% object@code
+
+  if (
+    any(a_is_shared_code) &&
+    (isFALSE(a_is_shared_code[1]) || isFALSE(b_is_shared_code[1]))
+  ) {
+    stop(
+      "`object` and `object2` can't be joined if one of the objects contains extra code before ",
+      "code shared by both."
+    )
+  } else if (
+    any(a_is_shared_code) && any(b_is_shared_code) &&
+    any(!a_is_shared_code) && any(!b_is_shared_code) &&
+    any(is_overwritten)) {
+    stop(
+      "`object` and `object2` have an extra code and environments contains modified object(s) of the same name ",
+      "Following object(s) have been modified:\n - ",
+      paste(common_names[is_overwritten], collapse = "\n - ")
+    )
+  }
+
+  # join expressions
+  # - remove duplicated code from object2 and append to the object1
+  unique_expr <- object2@code[!(object2@code %in% object@code & names(object2@code) %in% names(object@code))]
+  # - in case object@code and object2@code contains extra code with the same name - code2 needs to be renamed
+  object@code <- .keep_code_name_unique(object@code, unique_expr)
+
+  # insert (and overwrite) objects from env2 to env
+  object@env <- .copy_env(object@env)
+  lapply(ls(object2@env), function(x) assign(x, get(x, object2@env), object@env))
+
+  object
```

Comment:
how about the following?
https://stackoverflow.com/questions/26057400/r-how-do-you-merge-combine-two-environments

we would omit copying objects from one env to another which could be time consuming and memory inefficient

Reply:
very good suggestion!

Reply:
I got one thing in mind. Imagine multiple reactives. 
- We start on top with `q` and we have two independent reactive outputs `q1`, `q2` which will have `q` as parent. 
- Then `q3` depends on `q2` and links another environment
- on the bottom if we join `q1` and `q3` we will need to figure out again all parents 

![image](https://user-images.githubusercontent.com/6959016/180783081-782205a8-9516-42aa-a1f3-30efb6121c8d.png)

Reply:
What I mean above is that in `q3` we don't want to link last environment to `q1` but we need to find first environment after the last shared environment. 
On the other hand "copying" objects is made currently by reference, which means that we only make a new binding but we don't change their pointers- they will change if we modify the object (for example filter or add column)

Reply:
I don't quite get below statements:

> two independent reactive outputs q1, q2 which will have q as parent

> Then q3 depends on q2 and links another environment

> we need to find first environment after the last shared environment

The problem here is as follows:
Does quosure preserve its parent env after evaluation?

This stems from the different understanding of eval method. My naive understanding could be summarised into the following:

```
eval <- function(obj, code) {
  ... # append code 
  obj@env <- with(obj@env, eval(code, obj@env))
}
```

This does not preserve any parent-child relationship of environments.

As a result - we are not able to find last shared env. 

Your example would be as follows:
```
(pseudocode)
q <- new_quosure()

q1 <- eval_code(q, x <- 1)
# q1@obj: env(x = 1)

q2 <- eval_code(q, y <- 2)
# q2@obj: env(y = 2)

q3 <- eval_code(q2, x <- 3)
# q3@obj: env(y = 2, x = 3)

qq <- join(q1, q3)
# qq@obj: env(y = 2, x = 3)
```

Reply:
It means that if you change the parent of `q3` by:
```r
parent(q3@env) <- q1@env
``` 

then link between `q3` and `q2` will be lost, because `q2` will no longer be a parent of `q3`. It means that objects created in `q2` will no longer be in the joined environment because a env-tree would look like following:
```r
qq <- q1 <- q3
```

While it was
```
qq -- q1
   \_ q2 -- q3
```

Reply:
OK I got your point now and you are right - reference based joins won't work out. But I still believe that the quosure should not have a non-emptyenv parent env. I modified my idea into the following:

```
# new quo
q0 <- list2env(list(x = 0), parent = emptyenv())
rlang::env_print(q0)

# eval x = 1 in empty quo
#q1 <- list2env(list(x = 1), parent = q0)
q1 <- list2env(list(x = 1), parent = emptyenv()); rlang::env_coalesce(q1, q0)
for (i in c(q1, rlang::env_parents(q1))) rlang::env_print(i)
stopifnot(get("x", q1) == 1)
stopifnot(inherits(try(get("y", q1), silent = TRUE), "try-error"))
stopifnot(lobstr::obj_addr(get("x", q1)) != lobstr::obj_addr(get("x", q0)))

# eval y = 2 in empty quo
#q2 <- list2env(list(y = 2), parent = q0)
q2 <- list2env(list(y = 2), parent = emptyenv()); rlang::env_coalesce(q2, q0)
for (i in c(q2, rlang::env_parents(q2))) rlang::env_print(i)
stopifnot(get("x", q2) == 0)
stopifnot(get("y", q2) == 2)
stopifnot(lobstr::obj_addr(get("x", q2)) == lobstr::obj_addr(get("x", q0)))

# eval x = 3 in q2
#q3 <- list2env(list(x = 3), parent = q2)
q3 <- list2env(list(x = 3), parent = emptyenv()); rlang::env_coalesce(q3, q2)
for (i in c(q3, rlang::env_parents(q3))) rlang::env_print(i)
stopifnot(get("x", q3) == 3)
stopifnot(get("y", q3) == 2)
stopifnot(lobstr::obj_addr(get("x", q3)) != lobstr::obj_addr(get("x", q0)))
stopifnot(lobstr::obj_addr(get("y", q2)) == lobstr::obj_addr(get("y", q2)))

# join q3 into a q1
#qq <- list2env(as.list(q3), parent = q1)
qq <- new.env(parent = emptyenv()); rlang::env_coalesce(qq, q3); rlang::env_coalesce(qq, q1)
for (i in c(qq, rlang::env_parents(qq))) rlang::env_print(i)
stopifnot(get("x", qq) == 3)
stopifnot(get("y", qq) == 2)
stopifnot(lobstr::obj_addr(get("x", qq)) == lobstr::obj_addr(get("x", q3)))
stopifnot(lobstr::obj_addr(get("y", qq)) == lobstr::obj_addr(get("y", q2)))
```

commented-out code is my old and incorrect understanding
non-commented-out code is my new idea that seems to work on this dummy example

Reply:
Ok, I see your point. By non-empty you probably mean `parent.env(.GlobalEnv)` because if we stem from `emptyenv` we won't have access to any of the packages namespaces. I will use `rlang::env_coalesce` instead of `for` loop to "merge" environments.

Reply:
We have to use parent.env(.GlobalEnv) and it have to be dynamic to take into account any library called internally.
To be more precious  parent.env(.GlobalEnv) provided to us all base packages. Later when we execute any code with library call we have to update such environment to once more be one level below the Globalenv so we have access to all needed attached packages.

Reply:
`parent.env(object@env) <- parent.env(.GlobalEnv)` is applied in `new_quosure` and in `eval_code`. It's important that `eval_code` keeps updates parent in each evaluation - then modification caused by `library()` calls doesn't influence quosure so much.